rasa_nlu/extractors/crf_entity_extractor.py

```diff
@@ -0,0 +1,241 @@
+                else:
+                    word = sentence[word_idx - 1 + i]
+                    prefix = prefixes[i]
+                    features = self.crf_features[i]
+                    for feature in features:
+                        # append each feature to a feature vector
+                        word_features.append(prefix + feature + ':' + self.function_dict[feature](word))
+            sentence_features.append(word_features)
+        return sentence_features
+
+    def _sentence_to_labels(self, sentence):
+        labels = []
+        for word in sentence:
+            labels.append(word[2])
+        return labels
+
+    def _from_json_to_crf(self, json_eg):
+            #takes the json examples and switches them to a format which crfsuite likes
+            doc = self.nlp(json_eg[0])
+            entity_offsets = json_eg[1]
+            gold = GoldParse(doc, entities=entity_offsets)
+            ents = map(lambda l: l[5], gold.orig_annot)
+            if self.BILOU_flag == False:
+                def ent_clean(entity):
+                    if entity.startswith('B-') or entity.startswith('I-') or entity.startswith('U-') or entity.startswith('L-'):
+                        return entity[2:]
+                    else:
+                        return entity
+            else:
+                ent_clean = lambda l: l
+
+            crf_format = [(doc[i].text, doc[i].tag_, ent_clean(ents[i])) for i in xrange(len(doc))]
+            return crf_format
+
+    def _train_model(self, df_train):
+        # train the crf tagger based on the training data
+        self.ent_tagger = pycrfsuite.Tagger()
+
+        X_train = [self._sentence_to_features(sent) for sent in df_train]
+        y_train = [self._sentence_to_labels(sent) for sent in df_train]
+        trainer = pycrfsuite.Trainer(verbose=False)
+
+        for xseq, yseq in zip(X_train, y_train):
+            trainer.append(xseq, yseq)
+
+        trainer.set_params({
+            'c1': 1.0,   # coefficient for L1 penalty
+            'c2': 1e-3,  # coefficient for L2 penalty
+            'max_iterations': 50,  # stop earlier
+
+            # include transitions that are possible, but not observed
+            'feature.possible_transitions': True
+            })
+        self.f = NamedTemporaryFile()
+        a = time.time()
+        trainer.train(self.f.name)
+        self.ent_tagger.open(self.f.name)
+
+    def _test_model(self, df_test):
+
+        X_test = [self._sentence_to_features(sent) for sent in df_test]
+        y_test = [self._sentence_to_labels(sent) for sent in df_test]
+        y_pred = [self.ent_tagger.tag(xseq) for xseq in X_test]
+        print(bio_classification_report(y_test, y_pred))
+        
+        from collections import Counter
+        info = self.ent_tagger.info()
+
+
+
+
+def bio_classification_report(y_true, y_pred):
+    # taken from https://github.com/scrapinghub/python-crfsuite/blob/master/examples/CoNLL%202002.ipynb
```

Comment:
since we don't distribute pycrfsuite we are all good there. It's only an issue with code that we make available